lib/filterx/func-set-timestamp.c

```diff
@@ -0,0 +1,173 @@
+{
+  FilterXFunctionSetTimestamp *self = (FilterXFunctionSetTimestamp *) s;
+
+  FilterXObject *datetime_obj = filterx_expr_eval(self->datetime_expr);
+  if (!datetime_obj)
+    {
+      filterx_eval_push_error("Failed to evaluate second argument. " FILTERX_FUNC_SET_TIMESTAMP_USAGE, s, NULL);
+      return NULL;
+    }
+
+  UnixTime datetime = UNIX_TIME_INIT;
+
+  if (!filterx_object_extract_datetime(datetime_obj, &datetime))
+    {
+      filterx_object_unref(datetime_obj);
+      filterx_eval_push_error("Second argument must be of datetime type. " FILTERX_FUNC_SET_TIMESTAMP_USAGE, s, NULL);
+      return NULL;
+    }
+
+  filterx_object_unref(datetime_obj);
+
+  FilterXEvalContext *context = filterx_eval_get_context();
+  LogMessage *msg = context->msg;
+
+  msg->timestamps[self->timestamp_idx] = datetime;
+
+  return filterx_boolean_new(TRUE);
+}
+
+static void
+_set_timestamp_free(FilterXExpr *s)
+{
+  FilterXFunctionSetTimestamp *self = (FilterXFunctionSetTimestamp *) s;
+
+  filterx_expr_unref(self->datetime_expr);
+  filterx_function_free_method(&self->super);
+}
+
+static const gchar *
+_extract_set_timestamp_idx_str(FilterXFunctionArgs *args, GError **error)
+{
+  const gchar *idx_str = filterx_function_args_get_literal_string(args, 1, NULL);
+  if (!idx_str)
+    {
+      g_set_error(error, FILTERX_FUNCTION_ERROR, FILTERX_FUNCTION_ERROR_CTOR_FAIL,
+                  "argument must be set: timestamp. " FILTERX_FUNC_SET_TIMESTAMP_USAGE);
+      return NULL;
+    }
+
+  return idx_str;
+}
+
+static FilterXExpr *
+_extract_set_timestamp_datetime_expr(FilterXFunctionArgs *args, GError **error)
+{
+  FilterXExpr *datetime_expr = filterx_function_args_get_expr(args, 0);
+  if (!datetime_expr)
+    {
+      g_set_error(error, FILTERX_FUNCTION_ERROR, FILTERX_FUNCTION_ERROR_CTOR_FAIL,
+                  "argument must be set: datetime. " FILTERX_FUNC_SET_TIMESTAMP_USAGE);
+      return NULL;
+    }
+
+  return datetime_expr;
+}
+
+static gboolean
+_extract_set_timestamp_args(FilterXFunctionSetTimestamp *self, FilterXFunctionArgs *args, GError **error)
+{
+  gsize len = filterx_function_args_len(args);
+
+  if (len != 2)
+    {
+      g_set_error(error, FILTERX_FUNCTION_ERROR, FILTERX_FUNCTION_ERROR_CTOR_FAIL,
+                  "invalid number of arguments. " FILTERX_FUNC_SET_TIMESTAMP_USAGE);
+      return FALSE;
+    }
+
+  self->datetime_expr = _extract_set_timestamp_datetime_expr(args, error);
+  if (!self->datetime_expr)
+    {
+      return FALSE;
+    }
+
+  gchar *idx_str = g_strdup(_extract_set_timestamp_idx_str(args, error));
+  if (!idx_str)
+    {
+      return FALSE;
+    }
+
+  if (g_str_equal(idx_str, "ts_stamp"))
+    {
+      self->timestamp_idx = LM_TS_STAMP;
+      return TRUE;
+    }
+  if (g_str_equal(idx_str, "ts_recvd"))
+    {
+      self->timestamp_idx = LM_TS_RECVD;
+      return TRUE;
+    }
+  if (g_str_equal(idx_str, "ts_processed"))
+    {
+      self->timestamp_idx = LM_TS_PROCESSED;
+      return TRUE;
+    }
```

Comment:
I think most of the time the user wants to set the `stamp` timestamp, maybe sometimes `recvd`. Setting `processed` is never needed IMO, that is a disk-buffer related strange timestamp.

`date-parser` defaults to `stamp` and can be set to `recvd`, but not `processed`, we could follow that.
(Using `log_msg_lookup_time_stamp_name()` would implicitly do so)

Doing so would make this a named/optional parameter, if we follow `date-parser()` it could be called `time_stamp`, but I think we can challenge that if we are not fully satisfied with it.